### PR TITLE
Deprecate order tax zone

### PR DESCRIPTION
**Pull request by mamhoff**
This PR is two commits which are extracted from #1537, so that the Giant Removal is hopefully easier to understand. 

First, its a fix to `promotion_handler/coupon_spec`, which had weird spec setup in places that led to hard-to-debug errors. Second, it's deprecating `Spree::Order#tax_zone`, which is conceptually a terrible idea and rather expensive to use.

With the deprecation, I can also safely drop extra behaviour to `Spree::Order#reload`. Yay!

### CODE REVIEW

File: core/spec/models/spree/promotion_handler/coupon_spec.rb [hunk 1]
```diff
@@ -252,80 +252,83 @@ def expect_adjustment_creation(adjustable:, promotion:, promotion_code:nil)
...
-            @rate1 = Spree::TaxRate.create(
-              amount: 0.10,
-              calculator: Spree::Calculator::DefaultTax.create,
-              tax_category: @category,
-              zone: @zone
-            )
-
-            @order = Spree::Order.create!(store: store)
-            allow(@order).to receive_messages coupon_code: "10off"
+            expect(order).to receive(:tax_address).at_least(:once).and_return(Spree::Tax::TaxLocation.new(country: zone.countries.first))
```
Comment (tvdeyen) [inline core/spec/models/spree/promotion_handler/coupon_spec.rb]: Haha. Great. Much better to read and understand 👏🏻

File: core/spec/models/spree/shipping_rate_spec.rb [hunk 2]
```diff
@@ -141,7 +141,9 @@
       end
 
       it "shows correct tax amount" do
```
Comment (tvdeyen) [inline core/spec/models/spree/shipping_rate_spec.rb]: I like to use `aggregate_failures` when using multiple expectations

**Comment by jrochkind** [comment]
Without `Spree::Order#tax_zone`, what's the right way to tell what tax zone an Order is in for calculating taxes?  Am I missing something?

**Comment by mamhoff** [comment]
@jrochkind An order's shipping address might actually match several zones with tax rates attached. It's far easier to check which zones have the state or country of the shipping address, and then get the tax rates for them. 

The way this worked before was: 

We run `Zone.match`, get an array of zones and arbitrarily select one of them and designate it the tax zone. THEN, we do a matching of zones against that zone, where we had to check whether the rates zones wholly contain the order's tax zone, and it's all really unnecessarily complex. If you feel like, watch my talk on the subject: https://www.youtube.com/watch?v=BRnY-BidYRM :)

**Comment by jhawthorn** [comment]
Thanks Martin. This is much more sensible.

Looking forward to the follow up removal of default_tax, which should simplify a little more and remove a query or two.

**Pull request merged**